src/reflect/scala/reflect/internal/util/Collections.scala

```diff
@@ -61,6 +61,19 @@ trait Collections {
     head
   }
 
+  final def sameElementsEquals(thiss: List[AnyRef], that: List[AnyRef]): Boolean = {
+    // Probably immutable, so check reference identity first (it's quick anyway)
+    (thiss eq that) || {
```

Comment:
The key difference is using `.equals` rather than `.==`. The latter is routed through `BoxesRuntime.equals` which needs to perform a few type tests on the operands before delegating to `equals` in order to paper over the differences between primitive and boxed numeric equality.